managers/memory/handle_mq.c

```diff
@@ -0,0 +1,301 @@
+#include <lego/list.h>
+#include <lego/msg_q.h>
+#include <lego/stat.h>
+#include <lego/slab.h>
+#include <lego/uaccess.h>
+#include <lego/files.h>
+#include <lego/syscalls.h>
+#include <lego/comp_common.h>
+#include <lego/kernel.h>
+#include <memory/thread_pool.h>
+
+unsigned int append(char* msg_data, unsigned int msg_size, struct list_head* name_nid_dict){
+	struct mc_msg_queue* tmp;
+
+	tmp = kmalloc(sizeof(struct mc_msg_queue), GFP_KERNEL);
+	if(!tmp){
+		printk("allocated wrong\n");
+		return MSG_RET_FAIL;
+	}
+
+	tmp->msg_data = kmalloc(sizeof(char)*(msg_size+1), GFP_KERNEL);
+	if(!tmp->msg_data){
+		kfree(tmp);
+		printk("allocated wrong\n");
+		return MSG_RET_FAIL;
+	}
+
+	strcpy(tmp->msg_data, msg_data);
+	tmp->msg_size = msg_size;
+
+	list_add_tail(&tmp->list, name_nid_dict);	
+
+	return MSG_RET_SUCCESS;
+}
+
+
+
+/* the msg data passed in should copy a new memory here, msg_data should point to a continous memory
+ */
+unsigned int pop(char* msg_data, int* msg_size, struct list_head* name_nid_dict){
+	struct list_head* next = name_nid_dict->next;
+
+	struct mc_msg_queue* item = list_entry(next, struct mc_msg_queue, list);
+
+	strcpy(msg_data, item->msg_data);
+	*msg_size = item->msg_size;
+
+	list_del(&item->list);
+
+	kfree(item->msg_data);	
+	kfree(item);
+	name_nid_dict = next;
+
+	return MSG_RET_SUCCESS;
+}
+
+
+/*
+ * print api, for debug
+ */ 
+void print(struct list_head* name_nid_dict){
+	struct mc_msg_queue *pos = NULL;
+	list_for_each_entry(pos, name_nid_dict, list){
+		printk(pos->msg_data);
+		printk(" ");
+		printk("message data: %d\n", pos->msg_size);
+	}
```

Comment:
need list thread synchronization like other functions

Reply:
this is just for debug. A simple print function. I can delete this one.